DIPs/DIP1xxx-RC.md

```diff
@@ -0,0 +1,232 @@
+
+## Description
+
+This DIP adds a second kind of argument/parameter, named. Alternative names is ordered and unordered arguments. Ordered arguments is the current arrangement requiring that all argument be passed in the same order of the parameters defining them. But with unordered arguments, they can appear in any order or if specified to be optional at the parameter side, not at all.
+
+Named arguments are not affected by the passing of unnamed arguments. It does not matter where they go. So ``func(1, 2, o:true)`` is the same as ``func(1, o:true, 2)`` or ``func(o:true, 1, 2)``.
+
+At the template side, if it is not a function and the template parameters does not have any unnamed parameters you may omit the curved brackets. So ``struct Foo(<T>) {}`` is equivalent to ``struct Foo<T> {}``.
+
+If a named argument does not have a default value, it must be assigned or it is an error.
+Named arguments can be in any order, they cannot have values depending on each other, default or otherwise.
+
+To get a tuple of all named parameter names for (template instances or non-templated) functions, the trait ``__traits(getNamedParameters, T)`` can be used.
+
+Named arguments may be specified on structs, classes, unions, template blocks and mixin templates. As well as functions and methods. When used with structs, classes, unions or template blocks named arguments may be accessed by their identifier. E.g.
+
+```D
+struct MyWorld<string Name> {
+}
+
+static assert(MyWorld!(Name:"Earth").Name) == "Earth");
+
+void hello(<string message="World!">) {
+	import std.stdio;
+	writeln("Hello ", message);
+}
+
+void main() {
+	// Will print "Hello Walter"
+	hello(message : "Walter");
+}
+```
+
+This feature is quite convenient because it means that variadic template arguments and variadic function arguments can come before named arguments. So:
+
+```D
+void func(T..., <alias FreeFunc>)(T t, <bool shouldFree=false>) {
+}
+
+void abc() {
+	func!(FreeFunc : someRandomFunc)(1, 2, 3, shouldFree : true);
+}
+```
+
+Is in fact valid.
+
+For convenience at the definition side, you may combine named arguments together or keep them seperate, it is up to you.
+
+```D
+struct TheWorld<string Name, Type> {
+}
+
+alias YourWorld = TheWorld!(Name:"Here", Type:size_t);
+
+void goodies(T, <T t>, <U:class=Object>)(string text) {
+}
+
+alias myGoodies = goodies!(int, t:8);
+alias myOtherGoodies = goodies!(U:Exception, string, t : "hi!");
+```
+
+Any symbol that has named arguments, the named arguments are not considered for overload resolution. This puts a requirement on the unnamed arguments being unique and easily resolved.
+
+### Grammar changes
+
+```diff
+TemplateParameters:
++    < NamedTemplateParameterList|opt >
+
+TemplateParameter:
++    < NamedTemplateParameterList|opt >

```

Comment:
Bad idea. < and > are binary operators so this is likely to cause confusion. A short UDA before a named argument works better.

Reply:
I've read the parser, it won't be a problem.

Reply:
I meant from programmer viewpoint

Reply:
Well, it is a matter of taste. I think we should wait for more opinions on this.

Reply:
The important thing is that the syntax is consistent and can be guessed as to where else it can be used. But yes, I am in agreement, lets wait.